Handle nans via `is_valid_value()`

Krebrov001 opened the pull request:
- Relates to #2252
- Fixes #2252 

Handle nans via `is_valid_value()`

Krebrov001 committed:
Add the `is_valid_value()` checker, **both** C and C++ interfaces

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Add the `PlatformIOTest.is_valid_value`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Add the documentation for the `is_valid_value()` checker

- For both the C and C++ interfaces.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

dannosliwcd commented on service/docs/source/GEOPM_CXX_MAN_PlatformIO.3.rst:
```diff
@@ -132,6 +134,16 @@ Singleton Accessor
+
+
+``is_valid_value()``
+  In geopm, the values for signals and controls can be invalid due to errors.
```
```suggestion
  In geopm, the values for signals and controls can be invalid due to errors or absent sampling data.
```

dannosliwcd commented on service/src/PlatformIO.cpp:
```diff
@@ -1419,4 +1424,8 @@ extern "C" {
     }
+
+    int geopm_pio_check_valid_value(double value) {
+        return geopm::platform_io().is_valid_value(value) ? 0 : -1;
```
Should this return a value of `enum geopm_error_e`?

Krebrov001 replied:
`GEOPM_ERROR_INVALID`

dannosliwcd reviewed (commented):
This relates to #2252, but it does not fix #2252. The fix for that issue would either be to guarantee that `nan` never occurs in our floating point code, or to use a more strict floating point policy with our compiler.

The second option is less intrusive to us, but potentially impacts our users (who may rely on nan-related fast-math optimizations for their best-case performance). The hope is that the change you posted here is sufficient for those users that depend on fast math optimizations and also wish to use `geopm_pio` interfaces in their applications.

Run the unit test suite with everything built under a recent version of the `icpx` compiler to reproduce #2252. It would also be nice to see whether our assumption about this change being sufficient for users is true. In that case, we'd need to compile a version of GEOPM PIO with the fix, and link to it from a test application that is built with nan-related fast math optimizations enabled.

Krebrov001 committed:
Addressing review feedbacks for the `is_valid_value()` checker

- make it return a `GEOPM_ERROR_INVALID` upon error, instead of the -1

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Add new integration `test_invalid_values`

- This test checks if a test program compiled with fast FP math
  can successfully use the geopm functions compiled with strict FP math.
- Some edits to the Makefile and configure file to integrate this test.
- Added a new m4 file for the `AX_CHECK_COMPILE_FLAG`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Modify the `MANIFEST.EXAMPT`

- Make the `test_license.sh` pass.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo commented on m4/ax_check_compile_flag.m4:
```diff
@@ -0,0 +1,53 @@
+#   NOTE: Implementation based on AX_CFLAGS_GCC_OPTION. Please keep this
+#   macro in sync with AX_CHECK_{PREPROC,LINK}_FLAG.
+#
+# LICENSE
+#
+#   Copyright (c) 2008 Guido U. Draheim <guidod@gmx.de>
+#   Copyright (c) 2011 Maarten Bosmans <mkbosmans@gmail.com>
+#
+#   Copying and distribution of this file, with or without modification, are
+#   permitted in any medium without royalty provided the copyright notice
+#   and this notice are preserved.  This file is offered as-is, without any
+#   warranty.
```
We will have to add this header to geopm/copying_headers and list this file in the associated MANIFEST.

cmcantalupo commented on service/docs/source/GEOPM_CXX_MAN_PlatformIO.3.rst:
```diff
@@ -132,6 +134,16 @@ Singleton Accessor
+Static Class Methods
+--------------------
+
+
```
```suggestion
```

cmcantalupo commented on service/docs/source/GEOPM_CXX_MAN_PlatformIO.3.rst:
```diff
@@ -132,6 +134,16 @@ Singleton Accessor
+  This function determines if the current *value* of the signal or control is
+  a valid value. It returns true if the *value* is valid, and false if the *value*
+  is not valid.
+
```
```suggestion


```

cmcantalupo commented on service/docs/source/geopm_pio.3.rst:
```diff
@@ -96,6 +96,8 @@ Link with ``-lgeopm`` **(MPI)** or ``-lgeopmpolicy`` **(non-MPI)**
        int geopm_pio_reset(void);
 
+       int geopm_pio_check_valid_value(double value);
+
```
```suggestion


```

cmcantalupo commented on service/docs/source/geopm_pio.3.rst:
```diff
@@ -210,14 +212,20 @@ Inspection Functions
+  released/deleted and reconstructed.  As a result, any signals and
+  controls that had been pushed will be cleared, any batch servers
+  that had been started will be stopped, and all registered IOGroups
+  will be reset.  **NOTE: the reset only applies to the Service
+  PlatformIO instance and does not affect the PlatformIO instance
+  managed by the GEOPM HPC runtime.**
```
This comment is completely unrelated to this pull request, but this Note is misleading.  The reset applies to all use cases of PlatformIO with the one exception of the GEOPM HPC Runtime use case.  In the GEOPM HPC Runtime usecase, a different singleton access function is used, so this reset does not apply to the HPC Runtime PlatformIO instance.

cmcantalupo commented on service/docs/source/geopm_pio.3.rst:
```diff
@@ -210,14 +212,20 @@ Inspection Functions
+  managed by the GEOPM HPC runtime.**
+
+``geopm_pio_check_valid_value()``
+  In geopm, the values for signals and controls can be invalid due to errors.
```
```suggestion
  In geopm, signals may be unavailable due to missing information.  In these cases
  where the requested signal temporarily unavailable an invalid value is returned.
 ```

cmcantalupo commented on service/docs/source/geopm_pio.3.rst:
```diff
@@ -210,14 +212,20 @@ Inspection Functions
+
+``geopm_pio_check_valid_value()``
+  In geopm, the values for signals and controls can be invalid due to errors.
+  This function determines if the current *value* of the signal or control is
```
```suggestion
  This function determines if the current *value* of the signal is
```

cmcantalupo committed:
Add new manifest for lx_compile_flag.m4 macro file

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -1419,4 +1424,8 @@ extern "C" {
         geopm::platform_io_reset();
     }
+
+    int geopm_pio_check_valid_value(double value) {
```
```suggestion
    int geopm_pio_check_valid_value(double value)
    {
```

Krebrov001 committed:
Addressing review comments

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

avilcheslopez commented:
Awesome! I see this is passing the build now. Let me look into getting approval for the license being added before we merge this in.

dannosliwcd commented:
BTW, this also fixes the linked issue now that you also added the new build flags

avilcheslopez committed:
Add ax_check_compile_flag.m4 to dist.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

dannosliwcd commented on service/configure.ac:
```diff
@@ -321,8 +321,12 @@ AC_COMPILE_IFELSE([
 
-[EXTRA_CFLAGS="$EXTRA_CFLAGS -Wall -Werror"]
-[EXTRA_CXXFLAGS="$EXTRA_CXXFLAGS -Wall -Werror"]
+m4_include([m4/ax_check_compile_flag.m4])
+AX_CHECK_COMPILE_FLAG([-fp-model="fast"], [FASTMATH="-fp-model=fast"], [FASTMATH="-ffast-math"], [], [])
+AX_CHECK_COMPILE_FLAG([-fp-model="precise"], [NO_FASTMATH="-fp-model=precise"], [NO_FASTMATH="-fno-fast-math"], [], [])
+
+[EXTRA_CFLAGS="$EXTRA_CFLAGS -Wall -Werror $NO_FASTMATH"]
+[EXTRA_CXXFLAGS="$EXTRA_CXXFLAGS -Wall -Werror $NO_FASTMATH"]
```
Can you verify that [this case](https://github.com/geopm/geopm/issues/2252#issuecomment-1213135326) from #2252 (monitor agent + geopmbench + icpx) produces a valid `geopm.report` file with this PR's current revision?

If that test passes, then everything LGTM. If that test fails, then I think these changes may be needed in the HPC Runtime's configure.ac as well.

avilcheslopez commented:
> This relates to #2252, but it does not fix #2252. The fix for that issue would either be to guarantee that `nan` never occurs in our floating point code, or to use a more strict floating point policy with our compiler.
> 
> The second option is less intrusive to us, but potentially impacts our users (who may rely on nan-related fast-math optimizations for their best-case performance). The hope is that the change you posted here is sufficient for those users that depend on fast math optimizations and also wish to use `geopm_pio` interfaces in their applications.
> 
> Run the unit test suite with everything built under a recent version of the `icpx` compiler to reproduce #2252. It would also be nice to see whether our assumption about this change being sufficient for users is true. In that case, we'd need to compile a version of GEOPM PIO with the fix, and link to it from a test application that is built with nan-related fast math optimizations enabled.

Helpful steps provided by Daniel:
- [ ] Build & run the GEOPM HPC runtime from dev, using icpx and icx compilers. geopmbench should be an executable output of this build.
- [ ] Confirm that you can run geopmlaunch with geopmbench (e.g., single server, two MPI processes, default geopmbench configuration with no input args)
- [ ] Confirm that the generated geopm.report file replicates the zeroed app totals shown in the previous comment on this issue.
- [ ] Re-build GEOPM with the fix in #2613, still using icpx and icx compilers. Re-run geopmbench under geopmlaunch.
- [ ] Confirm that the new geopm.report correctly displays non-zero totals.

avilcheslopez committed:
Add NAN fix in runtime's configure script as well

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>
